app/controllers/supplementary/DocumentsProducedController.scala

```diff
@@ -34,29 +42,129 @@ class DocumentsProducedController @Inject()(
+      .map(_.getOrElse(DocumentsProducedData(Seq())))
+
+    cachedData.flatMap { cache =>
+      boundForm
+        .fold(
+          (formWithErrors: Form[DocumentsProduced]) =>
+            Future.successful(BadRequest(documents_produced(appConfig, formWithErrors, cache.documents))),
+          validForm =>
+            actionTypeOpt match {
+              case Some(Add)             => addItem(validForm, cache)
+              case Some(SaveAndContinue) => saveAndContinue(validForm, cache)
+              case Some(Remove(values))  => removeItem(retrieveItem(Json.parse(values.headOption.get)), cache)
+              case _                     => errorHandler.displayErrorPage()
+          }
+        )
+    }
+  }
+
+  private def saveAndContinue(
+    userInput: DocumentsProduced,
+    cachedData: DocumentsProducedData
+  )(implicit request: AuthenticatedRequest[_], hc: HeaderCarrier): Future[Result] =
+    (userInput, cachedData.documents) match {
+      case (document, Seq())     => handleSaveAndContinueEmptyCache(document)
+      case (document, documents) => handleSaveAndContinueCache(document, documents)
+    }
+
+  private def handleSaveAndContinueCache(document: DocumentsProduced, documents: Seq[DocumentsProduced])(
+    implicit request: AuthenticatedRequest[_]
+  ) =
+    document match {
+      case _ if documents.length >= maxNumberOfItems =>
+        handleErrorPage(Seq(("", "supplementary.addDocument.maximumAmount.error")), document, documents)
+
+      case _ if documents.contains(document) =>
+        handleErrorPage(Seq(("", "supplementary.addDocument.duplicated")), document, documents)
+
+      case _ if document.isDefined =>
+        val updatedCache = DocumentsProducedData(documents :+ document)
+
+        customsCacheService.cache[DocumentsProducedData](supplementaryCacheId, formId, updatedCache).map { _ =>
+          Redirect(controllers.supplementary.routes.SummaryPageController.displayPage())
+        }
+
+      case _ if !document.isDefined =>
+        Future.successful(Redirect(controllers.supplementary.routes.SummaryPageController.displayPage()))
+    }
+
+  private def handleSaveAndContinueEmptyCache(document: DocumentsProduced)(implicit request: AuthenticatedRequest[_]) =
+    document match {
+      case _ if document.isDefined =>
+        val updateCache = DocumentsProducedData(Seq(document))
+
+        customsCacheService.cache[DocumentsProducedData](supplementaryCacheId(), formId, updateCache).map { _ =>
+          Redirect(controllers.supplementary.routes.SummaryPageController.displayPage())
         }
-      )
+
+      case _ if !document.isDefined =>
+        Future.successful(Redirect(controllers.supplementary.routes.SummaryPageController.displayPage()))
+    }
+
+  private def addItem(
+    userInput: DocumentsProduced,
+    cachedData: DocumentsProducedData
+  )(implicit request: AuthenticatedRequest[_], hc: HeaderCarrier): Future[Result] =
+    (userInput, cachedData.documents) match {
```

Comment:
missing case when document is not defined, different solution in previous comment